runtime/vm/createramclass.cpp

```diff
@@ -2064,6 +2064,24 @@ checkFlattenableFieldValueClasses(J9VMThread *currentThread, J9ClassLoader *clas
 		const U_32 modifiers = field->modifiers;
 		J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(field);
 		U_8 *signatureChars = J9UTF8_DATA(signature);
+
+		if (J9_ARE_ALL_BITS_SET(modifiers, J9FieldFlagIsNullRestricted)) {
```

Comment:
During class loading, we should check non-static null-restricted fields and throw `IncompatibleClassChangeError` when necessary. This check can be moved into `if (J9_ARE_NO_BITS_SET(modifiers, J9AccStatic))` a few lines below. Probably you can do something like:

```
if (J9_ARE_NO_BITS_SET(modifiers, J9AccStatic)) {
	if (('Q' == signatureChars[0])
             || J9_ARE_ALL_BITS_SET(modifiers, J9FieldFlagIsNullRestricted)
        ) {
```

For static null-restricted fields, the `IncompatibleClassChangeError` should be thrown during preparation. The checks should be added inside:
https://github.com/eclipse-openj9/openj9/blob/080f0db81157518a64831ee0c7b55ffeb6547d62/runtime/vm/ClassInitialization.cpp#L591-L643